content/explugins/on.md

```diff
@@ -0,0 +1,66 @@
++++
+title = "on"
+description = "*on* executes a command when a specified event is triggered."
+weight = 10
+tags = [  "plugin" , "on" ]
+categories = [ "plugin", "external" ]
+date = "2018-01-22T07:53:19+01:00"
+repo = "https://github.com/mholt/caddy"
+home = "https://caddyserver.com/docs/onevent"
+enabled = "default"
++++
+
+## Description
+
+*on* executes a command when a specified event is triggered. This can be useful for preparing to
+ serve a site by running a script or starting a background process when the server starts, or for
+ stopping it when the server exits.
+
+Each command that is executed is blocking, unless you suffix the command with a space and &, which
+will cause the command to be run in the background. (Do not do this when the server is exiting, or
+the command may not finish before its parent process exits.) The output and error of the command go
+to standard outpu and standard error, respectively. There is no standard input.
+
+A command will only be executed once for each time it appears in the Corefile. In other words, even
+if this directive is shared by more than one zone, a command will only execute once per appearance
+in the Corefile.
+
+Note that commands scheduled for the shutdown event will not execute if CoreDNS is force-terminated,
+for example, by using a "Force Quit" feature provided by your operating system. However, a typical
+SIGINT (Ctrl+C) will allow the shutdown commands to execute.
+
```

Comment:
Another note. Each reload of corefile (e.g. during graceful restart by SIGUSR1) adds new event handlers (hooks) from the current corefile in addition to already existing handlers created from previous reloads of corefile. The existing handlers cannot be removed. So, be careful to not include to updated corefile the commands which have been already loaded before, when restarting the server with SIGUSR1. Otherwise, the commands will trigger several times.

Reply:
So a simple reload will make a command that listen for startup events, start twice? And three times after another reload? That sounds like a really bad bug....

Reply:
:astonished:

Reply:
Is there an upstream bug for this, cause that is seems hard to believe? And I probably drop the *on* plugin, i.e. not defaulting to on because of this.

Reply:
actually, the same problem exists in startupShutdown plugin
But I agree that the issues (in both plugin and doc) should be fixed in caddy project.
I haven't found any existing issues regarding these problems, so, maybe I will submit new one soon